### PR TITLE
More sanity checks for parallel replicas

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 78a9731734df5414d9fbabce1e9d88f4720ed99a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63910/78a9731734df5414d9fbabce1e9d88f4720ed99a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 1]
```diff
@@ -177,9 +178,17 @@ class ParallelReplicasReadingCoordinator::ImplInterface
     virtual ~ImplInterface() = default;
 
     virtual ParallelReadResponse handleRequest(ParallelReadRequest request) = 0;
-    virtual void handleInitialAllRangesAnnouncement(InitialAllRangesAnnouncement announcement) = 0;
+    virtual void doHandleInitialAllRangesAnnouncement(InitialAllRangesAnnouncement announcement) = 0;
     virtual void markReplicaAsUnavailable(size_t replica_number) = 0;
 
+    void handleInitialAllRangesAnnouncement(InitialAllRangesAnnouncement announcement)
+    {
+        if (++sent_initial_requests > replicas_count)
```
Comment (nickitat) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:186]: this check looks silly, but the reality is we had(ve) a bug in joins with new analyzer that is recognisable exactly by this madness.

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 2]
```diff
@@ -437,10 +445,9 @@ void DefaultCoordinator::handleInitialAllRangesAnnouncement(InitialAllRangesAnno
     ++stats[replica_num].number_of_requests;
     replica_status[replica_num].is_announcement_received = true;
 
-    ++sent_initial_requests;
     LOG_DEBUG(log, "Sent initial requests: {} Replicas count: {}", sent_initial_requests, replicas_count);
 
-    if (sent_initial_requests == replicas_count)
+    if (sent_initial_requests == replicas_count - unavailable_replicas_count)
```
Comment (nickitat) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:450]: it is unrelated fix I did along the way

**Review by devcrafter — APPROVED** [review]
